pytools/tag.py

```diff
@@ -0,0 +1,91 @@
+from __future__ import annotations
+from dataclasses import dataclass
+from typing import (Tuple, Any, FrozenSet)
+
+# {{{ dotted name
+
+
+class DottedName:
+    """
+    .. attribute:: name_parts
+
+        A tuple of strings, each of which is a valid
+        Python identifier. No name part may start with
+        a double underscore.
+
+    The name (at least morally) exists in the
+    name space defined by the Python module system.
+    It need not necessarily identify an importable
+    object.
+
+    .. automethod:: from_class
+    """
+
+    def __init__(self, name_parts: Tuple[str, ...]):
+        if len(name_parts) == 0:
+            raise ValueError("empty name parts")
+
+        for p in name_parts:
+            if not p.isidentifier():
+                raise ValueError(f"{p} is not a Python identifier")
+
+        self.name_parts = name_parts
+
+    @classmethod
+    def from_class(cls, argcls: Any) -> DottedName:
+        name_parts = tuple(
+                [str(part) for part in argcls.__module__.split(".")]
+                + [str(argcls.__name__)])
+        if not all(not npart.startswith("__") for npart in name_parts):
+            raise ValueError(f"some name parts of {'.'.join(name_parts)} "
+                             "start with double underscores")
+        return cls(name_parts)
+
+
+# }}}
+
+# {{{ tag
+
+tag_dataclass = dataclass(init=True, eq=True, frozen=True, repr=True)
```

Comment:
dataclasses are 3.7+ only. On 3.6 (and only there), `setup.py` should require https://pypi.org/project/dataclasses/.